### Background

Acute non-traumatic coma (aNTC) is a common paediatric presentation to hospitals in sub-Saharan Africa and is associated with a high risk of death and neurological sequelae. Acute bacterial meningitis (ABM) and cerebral malaria (CM) are often the common causes of aNTC. However, the cause for a considerable proportion of aNTC admissions remains unknown and is referred to as a coma of unknown cause (CUC). These groups of children have a higher risk of morbidity and mortality. Therefore, identifying the underlying causes, including infectious aetiology, is of great public health importance. Omics approaches have revolutionised the diagnosis and circumvented challenges associated with clinical and laboratory diagnosis. This study aimed to classify patients clinically categorised as CUC and identify infectious aetiologies among children admitted to Kilifi County Hospital on the Kenyan Coast using transcriptomics and RNA-based metagenomic next-generation sequencing (mNGS) approaches. 

### Materials and methods

This study used RNA sequencing to analyse cerebrospinal fluid samples from children classified as CUC (n= 72), ABM (n=15), and CM (n= 13). A transcript signature of differentially expressed genes (DEGs) between CM and ABM was used to identify ABM and CM cases within CUC that conventional diagnostic methods might have missed. The RNA-based mNGS approach was also used to identify any pathogens present among CUC cases.

### Results

534 DEGs between ABM and CM identified a subset of CM-like CUCs within the clinically diagnosed CUCs. Out of the total 72 clinically diagnosed CUC cases, 40,  (55%) had a transcriptome profile like that of children with CM. Notably, the genes upregulated in non-CM-like CUC were associated with T-cell function, suggesting possible viral infection, while those of CM-like CUC were associated with neurological function. Haemoglobin was significantly lower in CM-like CUC patients when compared to other CUC cases. The mNGS approach identified infectious pathogens among children categorised as CUC (48/72), 31 in the CM-like CUC group and 17 in the non-CM-like CUC group. There were (24/40) eukaryotic pathogens in the CM-like CUC group and (15/32) in the non-CM-like CUC group. Interestingly, there were (n=11) Plasmodium falciparum cases in the CM-like CUC group as compared to (n=6) in the non-CM-like CUC group. For viruses, there were (12/40) cases in the CM-like CUC group and (15/32) cases in the non-CM-like CUC group. The most common viral infections were Cytomegalovirus (n=6), Herpes simplex virus (n=7), and Lymphocryptovirus (n=9), while P. falciparum (n=14) and Aspergillus spp (n=18) were commonly identified eukaryotic pathogens.

### Conclusion

This study has demonstrated that a large proportion of children diagnosed with CUC likely do not have a bacterial aetiology. Further, the study has shown that CUC cases likely have coinfections between viral infections, Plasmodium falciparum or Aspergillus spp. , highlighting that RNA sequencing can potentially improve disease diagnosis. This is a small single-site study, and the results need to be verified in a larger cohort.  
Scanning
